client_test.go

```diff
@@ -117,20 +75,161 @@ func (e *customErrWithCause) Cause() error {
 	return e.cause
 }
 
-func TestCaptureExceptionShouldNotUseCauseIfCauseIsNil(t *testing.T) {
-	client, scope, transport := setupClientTest()
-	err := &customErrWithCause{cause: nil}
-	client.CaptureException(err, nil, scope)
-	assertEqual(t, transport.lastEvent.Exception[0].Type, "*sentry.customErrWithCause")
-	assertEqual(t, transport.lastEvent.Exception[0].Value, "err")
+type captureExceptionTestGroup struct {
+	name  string
+	tests []captureExceptionTest
 }
 
-func TestCaptureExceptionShouldUseCauseIfCauseIsNotNil(t *testing.T) {
-	client, scope, transport := setupClientTest()
-	err := &customErrWithCause{cause: &customErr{}}
-	client.CaptureException(err, nil, scope)
-	assertEqual(t, transport.lastEvent.Exception[0].Type, "*sentry.customErr")
-	assertEqual(t, transport.lastEvent.Exception[0].Value, "wat")
+type captureExceptionTest struct {
+	name string
+	err  error
+	want []Exception
+}
+
+func TestCaptureException(t *testing.T) {
+	basicTests := []captureExceptionTest{
+		{
+			name: "NilError",
+			err:  nil,
+			want: []Exception{
+				{
+					Type:       "sentry.usageError",
+					Value:      "CaptureException called with nil error",
+					Stacktrace: &Stacktrace{Frames: []Frame{}},
+				},
+			},
+		},
+		{
+			name: "SimpleError",
+			err:  errors.New("custom error"),
+			want: []Exception{
+				{
+					Type:       "*errors.errorString",
+					Value:      "custom error",
+					Stacktrace: &Stacktrace{Frames: []Frame{}},
+				},
+			},
+		},
+	}
+
+	errorChainTests := []captureExceptionTest{
+		{
+			name: "MostRecentErrorHasStack",
+			err:  pkgErrors.WithStack(&customErr{}),
+			want: []Exception{
+				{
+					Type:  "*sentry.customErr",
+					Value: "wat",
+					// No Stacktrace, because we can't tell where the error came
+					// from and because we have a stack trace in the most recent
+					// error in the chain.
+				},
+				{
+					Type:       "*errors.withStack",
+					Value:      "wat",
+					Stacktrace: &Stacktrace{Frames: []Frame{}},
+				},
+			},
+		},
+		{
+			name: "ChainWithNilCause",
+			err:  &customErrWithCause{cause: nil},
+			want: []Exception{
+				{
+					Type:       "*sentry.customErrWithCause",
+					Value:      "err",
+					Stacktrace: &Stacktrace{Frames: []Frame{}},
+				},
+			},
+		},
+		{
+			name: "ChainWithoutStacktrace",
+			err:  &customErrWithCause{cause: &customErr{}},
+			want: []Exception{
+				{
+					Type:  "*sentry.customErr",
+					Value: "wat",
+				},
+				{
+					Type:       "*sentry.customErrWithCause",
+					Value:      "err",
+					Stacktrace: &Stacktrace{Frames: []Frame{}},
+				},
+			},
+		},
+	}
+
+	tests := []captureExceptionTestGroup{
+		{
+			name:  "Basic",
+			tests: basicTests,
+		},
+		{
+			name:  "ErrorChain",
+			tests: errorChainTests,
+		},
+	}
+
+	for _, grp := range tests {
+		for _, tt := range grp.tests {
+			tt := tt
```

Comment:
Is it for closure scoping purpose?

Reply:
Yes. Because `tt` is in the scope of the subtest, having a new variable for each test prevents running into https://github.com/golang/go/wiki/CommonMistakes and allows us to run tests in parallel if we wanted to.

The observed behavior in the current code base today would be the same without this line, but I like to have it to help whomever will maintain this later.

It is easy to imagine some contribution to make tests run in parallel (not a priority today because they are pretty fast anyway), the tests would be green and we would not realize that we're only running the last test case over and over...